[libc][NFC] Adjust use of off_t internally

mikhailramalho opened the pull request:
This patch includes changes related to the use of off_t in libc,
targeted at 32-bit systems: in several places, the offset is used either
as a long or an off_t (64-bit signed int), but in 32-bit systems a long
type is only 32 bits long.

1. Fix a warning in mmap where a long offset is expected, but we were
   passing an off_t. A static_cast and a comment were added to explain
   that we know we are ignoring the upper 32-bit of the off_t in 32-bit
   systems.
2. The code in pread and pwrite was slightly improved to remove a
   #ifdef LIBC_TARGET_ARCH_IS_RISCV32; we are using an if constexpr now.
3. The Linux file operations were changed to use off_t instead of a long
   where applicable. No changes were made to the standard API, e.g.,
   ftell returns the offset as an int so we added a static_cast and a
   comment explaining that this will cause a loss of integer precision
   in 32-bit systems.

michaelrj-google reviewed (approved):
LGTM

gchatelet commented on libc/src/unistd/linux/pread.cpp:
```diff
@@ -19,15 +19,18 @@ namespace LIBC_NAMESPACE {
+  if constexpr (sizeof(off_t) == 8 && sizeof(size_t) == 4) {
+    // This is a 32-bit system with a 64-bit offset.
+    long offset_low = static_cast<long>(offset);
+    long offset_high = static_cast<long>(offset >> 32);
```
Bystander comment, is this supposed to split offset 8 bytes into two 4 bytes ?

If so I find this code suspicious for several reasons:
 - why test `sizeof(size_t) == 4` if we're actually using the `long` type in the syscall?
 - to get the low bits I would expect to see a mask, I see a cast and so the semantic seems unclear,
 - what happens to the sign bit in case `offset` is negative?

I'm not saying the code is wrong, but it's not obviously correct either.

mikhailramalho replied:
> Bystander comment, is this supposed to split offset 8 bytes into two 4 bytes ?
> 
> If so I find this code suspicious for several reasons:
> 
> * why test `sizeof(size_t) == 4` if we're actually using the `long` type in the syscall?

This checks only if we are running in a 32-bit system.

> * to get the low bits I would expect to see a mask, I see a cast and so the semantic seems unclear,

The cast will assign the lowest bits of offset to offset_low... It should work like it was before `(long)offset`. If we only use a mask and pass offset to `syscall_impl`, we'll get a warning about the loss of precision, that's why I've added the explicit cast.

> * what happens to the sign bit in case `offset` is negative?

AFAIK, a right shift of negative numbers is implementation-defined, but I assume the compiler will generate an arithmetic right shift here, so the sign bit will be preserved.

> 
> I'm not saying the code is wrong, but it's not obviously correct either.

gchatelet commented on libc/src/unistd/linux/pwrite.cpp:
```diff
@@ -19,15 +19,19 @@ namespace LIBC_NAMESPACE {
+    // This is a 32-bit system with a 64-bit offset.
+    long offset_low = static_cast<long>(offset);
+    long offset_high = static_cast<long>(offset >> 32);
+    ret = LIBC_NAMESPACE::syscall_impl<ssize_t>(SYS_pwrite64, fd, buf, count,
```
same here

gchatelet commented on libc/src/unistd/linux/pwrite.cpp:
```diff
@@ -19,15 +19,19 @@ namespace LIBC_NAMESPACE {
+  if constexpr (sizeof(off_t) == 8 && sizeof(size_t) == 4) {
+    // This is a 32-bit system with a 64-bit offset.
+    long offset_low = static_cast<long>(offset);
+    long offset_high = static_cast<long>(offset >> 32);
```
same here

gchatelet reviewed (changes requested):
I'd like to see proof that the split is correct, since it's going to come up again I think a function + tests would get us covered.

mikhailramalho commented:
> I'd like to see proof that the split is correct, since it's going to come up again I think a function + tests would get us covered.

Do you mean adding more regressions to the code base?

Currently, the pread/pwrite changes are tests by `pread_pwrite_test.cpp`, while the file changes are tested by several tests already.

I tested it locally for rv32 and this patch doesn't introduce any regressions. I don't think the rv32 buildbot will test this for now, as it needs `LLVM_LIBC_FULL_BUILD=ON` but I plan to enable it soon.

mikhailramalho committed:
[libc][NFC] Adjust use of off_t internally

This patch includes changes to the use of off_t in libc,
targeted at 32-bit systems: in several places, the offset is used either
as a long or an off_t (64-bit signed int), but in 32-bit systems a long
type is only 32 bits long.

1. Fix a warning in mmap where a long offset is expected, but we were
   passing an off_t. A static_cast and a comment were added to explain
   that we know we are ignoring the upper 32-bit of the off_t in 32-bit
   systems.
2. The code in pread and pwrite was slightly improved to remove a
   #ifdef LIBC_TARGET_ARCH_IS_RISCV32; we are using an if constexpr now.
3. The Linux file operations were changed to use off_t instead of a long
   where applicable. No changes were made to the standard API, e.g.,
   ftell returns the offset as an int so we added a static_cast and a
   comment explaining that this will cause a loss of integer precision
   in 32-bit systems.
